File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        Held = [0] * n
        Sold = [0] * n
        Free = [0] * n
        Held[0] = - prices[0]

        for i in range(1, n):
            Held[i] = max(Held[i - 1], Free[i - 1] - prices[i])
            Sold[i] = Held[i - 1] + prices[i]
            Free[i] = max(Free[i - 1], Sold[i - 1])
        
        return max(Held[n - 1], Sold[n - 1], Free[n - 1])




        

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int curProfit = 0;
        int buyDay = 0;
        int sellDay = 1;

        while (sellDay < prices.length) {
            curProfit = prices[sellDay] - prices[buyDay];
            if (curProfit > maxProfit) {
                maxProfit = curProfit;
            }
            else if (curProfit < 0) {
                buyDay = sellDay;
            }
            sellDay = sellDay + 1;
        }

        return maxProfit;
    }
}
